pkg/frontend/frontend.go

```diff
@@ -152,6 +166,74 @@ func NewFrontend(ctx context.Context,
 	return f, nil
 }
 
+// NewKubernetesClient is a wrapper for restconfig.RestConfig and kubernetes.NewForConfig
+// which returns (kubernetes.Interface, error) instead of (*kubernetes.Clientset, error),
+// which allows it to be used as a kubernetesClientFactory
+func NewKubernetesClient(args ...interface{}) (kubernetes.Interface, error) {
+	if len(args) != 2 {
+		return nil, errors.New("Required arguments: (env.Interface, *api.OpenShiftCluster)")
+	}
+	env, ok := args[0].(env.Interface)
+	if !ok {
+		return nil, errors.New("First argument must be of type env.Interface")
+
+	}
+	oc, ok := args[1].(*api.OpenShiftCluster)
+	if !ok {
+		return nil, errors.New("Second argument must be of type *api.OpenShiftCluster")
+	}
+	restConfig, err := restconfig.RestConfig(env, oc)
+	if err != nil {
+		return nil, err
+	}
+	return kubernetes.NewForConfig(restConfig)
+}
+
+// Option allows frontend fields to be overridden for unit testing
+type Option func(f *frontend)
+
+// WithKubeActionsFactory sets a KubeActions factory in frontend
+func WithKubeActionsFactory(factory kubeActionsFactory) Option {
+	return func(f *frontend) {
+		f.kubeActionsFactory = factory
+	}
+}
+
+// WithResourcesClientFactory sets a ResourcesClient factory in frontend
+func WithResourcesClientFactory(factory resourcesClientFactory) Option {
+	return func(f *frontend) {
+		f.resourcesClientFactory = factory
+	}
+}
+
+// WithVMClientFactory sets a VirtualMachinesClient factory in frontend
+func WithVMClientFactory(factory vmClientFactory) Option {
+	return func(f *frontend) {
+		f.vmClientFactory = factory
+	}
+}
+
+// WithVNetClientFactory sets a VirtualNetworksClient factory in frontend
+func WithVNetClientFactory(factory vNetClientFactory) Option {
+	return func(f *frontend) {
+		f.vNetClientFactory = factory
+	}
+}
+
+// WithAccountsClientFactory sets an AccountsClient factory in frontend
+func WithAccountsClientFactory(factory accountsClientFactory) Option {
+	return func(f *frontend) {
+		f.accountsClientFactory = factory
+	}
+}
+
+// WithKubernetesClientFactory sets a Kubernetes Clientset factory in frontend
+func WithKubernetesClientFactory(factory kubernetesClientFactory) Option {
+	return func(f *frontend) {
+		f.kubernetesClientFactory = factory
+	}
+}
```

Comment:
I wouldn't pass anything to these functions or else your `New` function is going to be as long as it is now during tests or any other environment.  

Typically in this scenario you would define the default in the New function and allow this to override when calling this for tests, etc. 

So if you wanted to pass `mock_metrics` to the `New` function you could put that function in the constructor and then the 
```go
  for _, opt := range options {
     opt(f)
}
```
would take care of the rest.

Reply:
`frontend.New` in `rp.go` is now much shorter, as are most of the other calls in unit tests. The default for these factories is the real thing. That default is overridden by the value of `factory` passed in the `With...` functions. We need to pass in the factory in unit tests so that we can expect calls to a previously created mock client in unit tests.